streaming.py: add option to pass metadata to returned messages

myNameIsPatrick opened the pull request:
Implements #7, adding metadata option to AlertBroker, which when set, allows protocol-specific metadata alongside message contents to be returned from the stream.

In order to avoid complications, I had to remove the message index being returned.

Additionally, filter now expects two arguments; the message and the metadata regardless of whether it will be returned.

I had thought about making this dependent on whether metadata was requested but that handling that internally would be more difficult and I would personally get confused on passing in a filter function that isn't consistent across options.

Additionally, I moved the handling of heartbeats happens further upstream to simplify the return type of messages from the stream, since metadata wasn't defined as the time the heartbeat was processed in _filtered_stream. This did end up making the downstream logic a bit simpler as a result.

The metadata structure is a named tuple, as shown below, to access properties by name. 

Example:
```
from genesis.streaming import open

with open(kafka_url, "r", start_at="beginning", format='json', metadata=True) as stream:
    for msg, meta in stream:
        print(f"{msg}, {meta}")
```
Output:
```
{'hey': 'you'}, Metadata(topic='gcn', partition=0, offset=0, timestamp=1584024596082, key=None)
```

Closes #7.

mjuric commented:
Looks good, save for one request -- can you add the equivalent of `idx` to the `Metadata` structure? From experience, looking at `idx` is very useful when monitoring/debugging the performance of e.g., filters (because idx increases for every message received, you can tell how many were filtered out).

Except for that, looks good to merge!

myNameIsPatrick commented:
No problem. I added the message index back in where the message metadata is generated and the error checking is done. As a result, you can increment the index after yielding the message at this stage since all the messages from the raw stream are fully processed before being sent to the mapper. I also moved the `validate_and_process` function to an `AlertBroker` method to have access to the index internally.
